src/classes/CW_AudienceCSVProcessHelper.cls

```diff
@@ -0,0 +1,406 @@
+   	 		}
+   	 		grupSheduleMap.put(grpId,rec);
+   	 		ListToIns.Add(rec);
+   		}
+   	}
+   	insert ListToIns;
+   }
+  
+   //This is complicated so we must process both to not make to much of shcedureRewards
+   //We have one shedule so we must find grup and reward(freebet)
+   private void createAudinceGroupScheduleReward(){
+ 
+   	//There a many so we must have coposit key grupId _FreeBetId (Campaign Reward)
+   	audSchedRewMap= new map<string,id>();
+   	for(CW_Audience_Group_Schedule_Reward__c rec :[SELECT Id,Audience_Group_Schedule__r.Audience_Group__c,Campaign_Reward__c FROM CW_Audience_Group_Schedule_Reward__c 
+   					WHERE Audience_Group_Schedule__r.Schedule__c =:CsvStatus.Schedule__c]){
+   		 audSchedRewMap.put(''+rec.Audience_Group_Schedule__r.Audience_Group__c+rec.Campaign_Reward__c,rec.ID);		
+   	}
+   	//GO for AllRecords and chek if we have all Keys
+   	List<CW_Audience_Group_Schedule_Reward__c> ToInsert= new List<CW_Audience_Group_Schedule_Reward__c>();
+   	for(CW_CSV_Stage_AudienceMembers__c rec:CsvStageList ){
+   		CW_Campaign_Reward__c campRew=campRewardMap.get(rec.Reward_Library__c);
+   		if(campRew==null){
+   				continue; // we propably should rise error as somthing wrong has hepen  -> we dont created sscheduled group in prev step !!!
+   			}
+   		String Key=''+rec.Audience_Group_ref__c+campRew.ID;
+   		//Regarding 27-05-2014 sprint 07  for control groups we can have rewards 
+   		//if(rec.Audience_Group_ref__r.Is_Control_group__c){
+   		//	continue ;// There no rewords for control groups.
+   		//}
+   		if(!audSchedRewMap.containskey(key)){
+   			CW_Audience_Group_Schedule_Reward__c newShedRew=new  CW_Audience_Group_Schedule_Reward__c();
+   			CW_Audience_Group_Schedule__c shedRec=grupSheduleMap.get(rec.Audience_Group_ref__c);
+   			
+   			if(shedRec==null){
+   				continue; // we propably should rise error as somthing wrong has hepen  -> we dont created sscheduled group in prev step !!!
+   			}
+   			
+   			newShedRew.Audience_Group_Schedule__c=shedRec.ID;
+   			newShedRew.Campaign_Reward__c=campRew.ID;
+   			ToInsert.Add(newShedRew);
+   			audSchedRewMap.put(key,null);//To not add twice
+   		}
+   		
+   	}
+  		insert ToInsert;
+  		//becouse we use fields from releted object we need to requery. mayby beter option will be ser __r=xxx and use upsert
+  		for(CW_Audience_Group_Schedule_Reward__c rec :[SELECT Id,Audience_Group_Schedule__r.Audience_Group__c,Campaign_Reward__c FROM CW_Audience_Group_Schedule_Reward__c 
+   					WHERE Audience_Group_Schedule__r.Schedule__c =:CsvStatus.Schedule__c]){
+   		 audSchedRewMap.put(''+rec.Audience_Group_Schedule__r.Audience_Group__c+rec.Campaign_Reward__c,rec.ID);		
+   	}
+ 
+   }
+   
+   
+   	
+   private void createAudinceMember(){
+ 		//  This is last stap we dont kreate globals  
+   	audMemberMap= new map<string,id>();
+   	for(CW_Audience_Group_Member__c rec :[SELECT Id,Audience_Group_LU__c,Contact__c FROM CW_Audience_Group_Member__c 
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.